cbor/shared/src/main/scala/fs2/data/cbor/high/internal/ValueParser.scala

```diff
@@ -0,0 +1,321 @@
+ */
+
+package fs2
+package data
+package cbor
+package high
+package internal
+
+import low.CborItem
+
+import scala.collection.mutable
+
+import java.lang.{Float => JFloat, Double => JDouble, Long => JLong}
+import scodec.bits.ByteVector
+
+object ValueParser {
+
+  private type Result[F[_], T] = (Chunk[CborItem], Int, Stream[F, CborItem], List[CborValue], T)
+
+  private def raise[F[_]](e: CborException, chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Nothing] =
+    Pull.output(Chunk.seq(chunkAcc.reverse)) >> Pull.raiseError(e)
+
+  private def parseArray[F[_]](chunk: Chunk[CborItem],
+                               idx: Int,
+                               rest: Stream[F, CborItem],
+                               size: Long,
+                               acc: mutable.ListBuffer[CborValue],
+                               chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue]] =
+    if (size == 0l) {
+      Pull.pure((chunk, idx, rest, chunkAcc, CborValue.Array(acc.result(), false)))
+    } else {
+      if (idx >= chunk.size) {
+        Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+          case Some((hd, tl)) => parseArray(hd, 0, tl, size, acc, Nil)
+          case None           => Pull.raiseError(new CborException("unexpected end of input"))
+        }
+      } else {
+        if (JLong.compareUnsigned(size, Int.MaxValue.toLong) > 0) {
+          raise(
+            new CborException(
+              s"array size is limited to max int (${Int.MaxValue}) elements but got ${JLong.toUnsignedString(size)}"),
+            chunkAcc)
+        } else {
+          parseValue(chunk, idx, rest, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc, value) =>
+            acc.append(value)
+            parseArray(chunk, idx, rest, size - 1, acc, chunkAcc)
+          }
+        }
+      }
+    }
+
+  private def parseIndefiniteArray[F[_]](chunk: Chunk[CborItem],
+                                         idx: Int,
+                                         rest: Stream[F, CborItem],
+                                         acc: mutable.ListBuffer[CborValue],
+                                         chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue]] =
+    if (idx >= chunk.size) {
+      Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+        case Some((hd, tl)) => parseIndefiniteArray(hd, 0, tl, acc, Nil)
+        case None           => Pull.raiseError(new CborException("unexpected end of input"))
+      }
+    } else {
+      chunk(idx) match {
+        case CborItem.Break =>
+          Pull.pure((chunk, idx + 1, rest, chunkAcc, CborValue.Array(acc.result(), true)))
+        case _ =>
+          if (acc.size == Int.MaxValue) {
+            raise(new CborException(s"array size is limited to max int (${Int.MaxValue}) elements"), chunkAcc)
```

Comment:
If this exception is ever thrown somewhere, we know fs2-data sees proper production usage!